src/btree/bt_debug.c

```diff
@@ -946,6 +946,40 @@ __wt_debug_cursor_page(void *cursor_arg, const char *ofile)
     return (ret);
 }
 
+/*
+ * __wt_debug_dump_tree --
+ *     Lock all the refs and dump the in-memory information.
+ */
+int
+__wt_debug_dump_tree(void *cursor_arg)
+{
+    WT_CURSOR_BTREE *cbt;
+    WT_DECL_RET;
+    WT_REF *ref;
+    WT_SESSION_IMPL *session;
+    uint8_t previous_state;
+
+    cbt = cursor_arg;
+    session = CUR2S(cursor_arg);
+
+    WT_INTL_FOREACH_BEGIN (session, cbt->ref->home, ref) {
+        previous_state = ref->state;
+        if (previous_state != WT_REF_MEM)
+            continue;
+        WT_REF_LOCK(session, ref, &previous_state);
+    }
+    WT_INTL_FOREACH_END;
+
+    WT_INTL_FOREACH_BEGIN (session, cbt->ref->home, ref) {
+        if (ref->state != WT_REF_LOCKED)
+            continue;
+        WT_WITH_BTREE(session, CUR2BT(cbt), ret = __wt_debug_page(session, NULL, ref, NULL));
+    }
+    WT_INTL_FOREACH_END;
+
+    return (ret);
+}
```

Comment:
Isn't this only going to lock the top-level WT_REF's? That is, we're not locking the WT_REF's down through the tree as we walk it, we're just locking the top page's WT_REFs.

Staring at this a little more, maybe the simplest approach is to change `__debug_page()` to lock/unlock the WT_REF all the time, that is, make it standard that the dump routines lock/unlock WT_REFs so we minimize the chance for dropping core. Then we can call `__wt_debug_tree_all` from the order-check code.

I know I suggested a separate locking pass, but I don't have any evidence it's needed or would make a difference, and it's going to require a lot more code.

Reply:
As a side comment -- maybe I'm missing something, but opening/closing the HS cursor in `__debug_page()` is going to be awfully expensive as we're dumping the tree. Maybe we should pull that out into the callers of `__debug_page()` so we only do it once for each dump?

@raviprakashgiri29, I'm sorry to keep changing the target on this work, and I haven't looked at this as closely as you have, so maybe I'm just wrong -- regardless, happy to talk this over with you if that would be useful.